inner_products/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+use algebra::{
+    bytes::ToBytes,
+    curves::PairingEngine,
+    fields::{Field, PrimeField},
+    groups::Group,
+};
+use std::{
+    error::Error as ErrorTrait,
+    fmt::{Display, Formatter, Result as FmtResult},
+    io::{Result as IoResult, Write},
+    marker::PhantomData,
+    ops::{Add, Mul, MulAssign},
+};
+
+pub type Error = Box<dyn ErrorTrait>;
+
+#[derive(Debug)]
+pub enum InnerProductError {
+    MessageLengthInvalid(usize, usize),
+}
+
+impl ErrorTrait for InnerProductError {
+    fn source(self: &Self) -> Option<&(dyn ErrorTrait + 'static)> {
+        None
+    }
+}
+
+impl Display for InnerProductError {
+    fn fmt(self: &Self, f: &mut Formatter<'_>) -> FmtResult {
+        let msg = match self {
+            InnerProductError::MessageLengthInvalid(left, right) => {
+                format!("left length, right length: {}, {}", left, right)
+            }
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+pub trait InnerProduct: Copy {
+    type LeftMessage;
+    type RightMessage;
+    type Output;
+
+    fn inner_product(
+        left: &[Self::LeftMessage],
+        right: &[Self::RightMessage],
+    ) -> Result<Self::Output, Error>;
+}
+
+#[derive(Copy, Clone)]
+pub struct PairingInnerProduct<P: PairingEngine> {
+    _pair: PhantomData<P>,
+}
+
+impl<P: PairingEngine> InnerProduct for PairingInnerProduct<P> {
+    type LeftMessage = P::G1Projective;
+    type RightMessage = P::G2Projective;
+    type Output = ExtensionFieldElement<P>;
+
+    fn inner_product(
+        left: &[Self::LeftMessage],
+        right: &[Self::RightMessage],
+    ) -> Result<Self::Output, Error> {
+        if left.len() != right.len() {
+            return Err(Box::new(InnerProductError::MessageLengthInvalid(
+                left.len(),
+                right.len(),
+            )));
+        };
+        Ok(ExtensionFieldElement(
+            left.iter()
+                .zip(right)
+                .map(|(v, a)| P::pairing(v.clone().into(), a.clone().into()))
+                .product(),
+        ))
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct MultiexponentiationInnerProduct<G: Group> {
+    _group: PhantomData<G>,
+}
+
+impl<G: Group> InnerProduct for MultiexponentiationInnerProduct<G> {
+    type LeftMessage = G;
+    type RightMessage = G::ScalarField;
+    type Output = G;
+
+    fn inner_product(
+        left: &[Self::LeftMessage],
+        right: &[Self::RightMessage],
+    ) -> Result<Self::Output, Error> {
+        if left.len() != right.len() {
+            return Err(Box::new(InnerProductError::MessageLengthInvalid(
+                left.len(),
+                right.len(),
+            )));
+        };
+        Ok(left.iter().zip(right).map(|(g, x)| g.mul(x)).sum())
```

Comment:
For larger inputs (len > 100), you might want to use something like https://github.com/scipr-lab/zexe/blob/38b6f6f2c6b9a6a1b5e8eb90b1287c98b06520d1/algebra-core/src/msm/variable_base.rs#L100

It should speed up this inner product massively.

Reply:
:+1: 

Had to change the generic parameter away from `Group` to a curve in order to use `VariableMSM::multi_scalar_mul`.